PagingWithNetworkSample/benchmark/src/androidTest/java/com/example/benchmark/BenchmarkActivity.kt

```diff
@@ -0,0 +1,60 @@
+package com.example.benchmark
+
+import android.os.Bundle
+import android.util.Log
+import androidx.appcompat.app.AppCompatActivity
+import androidx.paging.PageKeyedDataSource
+import androidx.paging.PagedList
+import com.android.example.paging.pagingwithnetwork.GlideApp
+import com.android.example.paging.pagingwithnetwork.reddit.repository.NetworkState
+import com.android.example.paging.pagingwithnetwork.reddit.ui.PostsAdapter
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPost
+import kotlinx.android.synthetic.main.activity_benchmark.*
+import java.util.concurrent.Executors
+
+class BenchmarkActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_benchmark)
+
+        val glide = GlideApp.with(this)
+        val adapter = PostsAdapter(glide) {}
+        list.adapter = adapter
+
+        val config = PagedList.Config.Builder()
+                .setInitialLoadSizeHint(5)
+                .setPageSize(5)
+                .build()
+
+        val pagedStrings: PagedList<RedditPost> = PagedList.Builder<Int, RedditPost>(MockDataSource(), config)
+                .setInitialKey(0)
+                .setFetchExecutor(Executors.newSingleThreadExecutor())
+                .setNotifyExecutor {
+                    runOnUiThread { it.run() }
+                }
+                .build()
+
+        adapter.submitList(pagedStrings)
+        adapter.setNetworkState(NetworkState.LOADED)
+    }
+}
+
+class MockDataSource : PageKeyedDataSource<Int, RedditPost>() {
+    override fun loadInitial(params: LoadInitialParams<Int>, callback: LoadInitialCallback<Int, RedditPost>) {
+        callback.onResult((0..5).map { generatePost() }.toList(), 1, 2)
+    }
+
+    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, RedditPost>) {
+        callback.onResult((0..5).map { generatePost() }.toList(), params.key + 1)
+    }
+
+    override fun loadBefore(params: LoadParams<Int>, callback: LoadCallback<Int, RedditPost>) {
+        callback.onResult((0..5).map { generatePost() }.toList(), params.key - 1)
+    }
+
+    private fun generatePost(): RedditPost {
+        val title = (0..10).map { (0..100).random() }.joinToString("")
```

Comment:
If you vary the length of these, or randomize them, does it affect the benchmark significantly?

One of the things we'll probably be measuring is layout cost, so randomly generating these strings to be similar to real data is useful in telling someone how the list will actually perform.

Android text layout does a lot of its work on a per-word basis. But if it's seen a word before recently, it gets to skip all of that.

That's why when benchmarking text, we need to be careful that we're not just benchmarking the fast every-word-is-cached case.

There's a sample in the platform code, frameworks/base:

https://github.com/aosp-mirror/platform_frameworks_base/blob/6bebb8418ceecf44d2af40033870f3aabacfe36e/tests/UiBench/src/com/android/test/uibench/TextUtils.java#L68

Which I've used to create fake text for recyclerview before. It intermixes a short static list of words that will always hit, with randomly generated sequences of characters. This is in uibench, a sort of cousin of the benchmarking library. What that does is allow you to generate text that will have a consistent hitrate in the layout cache.

Note: suggest leaving as is for now, and fixing it in a targeted follow up PR. Then you can also add comments about how to benchmark text.

Reply:
Wait - isn't a random string generated here everytime?

Reply:
Oh, yeah it is - maybe I was commenting about a previous version?